engines/cufile.c

```diff
@@ -0,0 +1,586 @@
+	size_t              total_mem;     /* size for cu_mem_ptr and junk_buf */
+	int                 logged;        /* bitmask of log messages that have
+										  been output, prevent flood */
+};
+
+struct fio_cufile_data {
+	CUfileDescr_t  cf_descr;
+	CUfileHandle_t cf_handle;
+};
+
+struct fio_option options[] = {
+	{
+		.name	= "cufile_gpu_ids",
+		.lname	= "cufile gpu ids",
+		.type	= FIO_OPT_STR_STORE,
+		.off1	= offsetof(struct cufile_options, gpu_ids),
+		.help	= "GPU IDs, one per job, separated by " GPU_ID_SEP,
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= "cufile_posixio",
+		.lname	= "cufile posixio",
+		.type	= FIO_OPT_BOOL,
+		.off1	= offsetof(struct cufile_options, use_posixio),
+		.help	= "Use POSIX I/O instead of cuFile API",
+		.def    = "0",
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= NULL,
+	},
+};
+
+static int running = 0;
+static int cufile_initialized = 0;
+static pthread_mutex_t running_lock = PTHREAD_MUTEX_INITIALIZER;
+
+#define check_cudaruntimecall(fn, rc)								\
+	do {															\
+		cudaError_t res = fn;										\
+		if (res != cudaSuccess) {									\
+			const char *str = cudaGetErrorName(res);				\
+			log_err("cuda runtime api call failed %s:%d : %s\n",	\
+					#fn, __LINE__, str);							\
+			rc = -1;												\
+		} else {													\
+			rc = 0;													\
+		}															\
+	} while(0)
+
+static const char* cuFileGetErrorString(CUfileError_t st)
+{
+	if (IS_CUFILE_ERR(st.err)) {
+		return cufileop_status_error(st.err);
+	}
+	return "unknown";
+}
+
+/**
+ * fio_cufile_init called in job process after fork() or pthread_create().
+ */
+static int fio_cufile_init(struct thread_data *td)
+{
+	struct cufile_options *o = td->eo;
+	char* cur = NULL;
+	char* pos = NULL;
+	CUfileError_t status;
+	int gpu_id = 0;
+	int i;
+	int initialized;
+	int rc;
+
+	pthread_mutex_lock(&running_lock);
+	if (running == 0) {
+		assert(cufile_initialized == 0);
+		if (!o->use_posixio) {
+			/* only open the driver if this is the first worker thread */
+			status = cuFileDriverOpen();
+			if (status.err != CU_FILE_SUCCESS) {
+				log_err("cuFileDriverOpen: %d:%s\n", status.err,
+						cuFileGetErrorString(status));
+			} else {
+				cufile_initialized = 1;
+			}
+		}
+	}
+	running++;
+	initialized = cufile_initialized;
+	pthread_mutex_unlock(&running_lock);
+
+	if (!o->use_posixio && !initialized) {
+		return 1;
+	}
+
+	if (o->gpu_ids) {
+		for (i = 0, cur = strtok_r(o->gpu_ids, GPU_ID_SEP, &pos);
+			 i <= td->subjob_number && cur;
+			 ++i, cur = strtok_r(NULL, GPU_ID_SEP, &pos)) {
+			gpu_id = atoi(cur);
+		}
+	}
+
+	o->my_gpu_id = gpu_id;
+
+	check_cudaruntimecall(cudaSetDevice(o->my_gpu_id), rc);
+	if (rc != 0) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static inline int fio_cufile_pre_write(struct thread_data *td,
+									   struct cufile_options *o,
+									   struct io_u *io_u,
+									   size_t gpu_offset)
+{
```

Comment:
Again I can tell if this will really line up in real life...

Reply:
I will adjust my editor settings and reformat the code.